plan/logical_plan_builder.go

```diff
@@ -668,31 +668,86 @@ func (b *planBuilder) buildProjection4Union(u *LogicalUnionAll) {
 }
 
 func (b *planBuilder) buildUnion(union *ast.UnionStmt) LogicalPlan {
-	u := LogicalUnionAll{}.init(b.ctx)
-	u.children = make([]LogicalPlan, len(union.SelectList.Selects))
-	for i, sel := range union.SelectList.Selects {
-		u.children[i] = b.buildSelect(sel)
-		if b.err != nil {
-			return nil
-		}
-		if u.children[i].Schema().Len() != u.children[0].Schema().Len() {
-			b.err = errors.New("The used SELECT statements have a different number of columns")
-			return nil
-		}
+	distinctSelectPlans, allSelectPlans := b.resolveUnionSelectPlans(union.SelectList.Selects)
+	if b.err != nil {
+		return nil
 	}
 
-	b.buildProjection4Union(u)
-	var p LogicalPlan = u
-	if union.Distinct {
-		p = b.buildDistinct(u, u.Schema().Len())
+	unionDistinctPlan := b.buildSubUnion(distinctSelectPlans)
+	if b.err != nil {
+		return nil
 	}
+
+	if unionDistinctPlan != nil {
+		unionDistinctPlan = b.buildDistinct(unionDistinctPlan, unionDistinctPlan.Schema().Len())
+	}
+
+	if unionDistinctPlan != nil && allSelectPlans != nil && len(allSelectPlans) != 0 {
+		allSelectPlans = append(allSelectPlans, unionDistinctPlan)
+	}
+
+	unionAllPlan := b.buildSubUnion(allSelectPlans)
+	if b.err != nil {
+		return nil
+	}
+
+	var unionPlan LogicalPlan
+	if unionAllPlan != nil {
+		unionPlan = unionAllPlan
+	} else if unionDistinctPlan != nil {
+		unionPlan = unionDistinctPlan
+	} else {
+		return nil
+	}
+
 	if union.OrderBy != nil {
-		p = b.buildSort(p, union.OrderBy.Items, nil)
+		unionPlan = b.buildSort(unionPlan, union.OrderBy.Items, nil)
 	}
 	if union.Limit != nil {
-		p = b.buildLimit(p, union.Limit)
+		unionPlan = b.buildLimit(unionPlan, union.Limit)
 	}
-	return p
+	return unionPlan
+}
+
+// resolveUnionSelectPlans to resolve union's select stmts to logical plans.
+// and divide result plans into "union-distinct" and "union-all" parts.
```

Comment:
s/ divide/ divides
and it seems, `divideUnionSelectPlans` is better than `resolveUnionSelectPlans` ^_^